cedar-example-use-cases/discord/policies.cedar

```diff
@@ -0,0 +1,32 @@
+permit (
+  principal,
+  action == Action::"SendMessage",
+  resource
+)
+when
+{
+  // the user's channel matches the resource
+  (principal.channel == resource) &&
```

Comment:
Is it really the case that Discord makes users specific to a channel?  Or is this a modeling artifact?

If the latter, can we model this more simply without the entity hierarchy and channel-specific users as follows.  Each user has an attribute `channelRoles : Set<{channel : Channel, role : String}>`.  Then, to check if the user has “SendMessage” role, we write:

```
permit (
  principal, 
  action == Action::”SendMessage”,
  resource 
 when {
   principal.channelRoles.contains({channel : resource, role : “SendMessage” } )
 }
``` 

Would this work?

Reply:
Yeah, you've pointed out the problem: users should not be channel-specific.

Unfortunately, your cedar rule won't work because it drops the `principal.channel_roles in Allow::"SendMessage"` line. What this rule needs is a map from channels to `ChannelRoles` so that we can perform this lookup. (see the comment in the `User` entity)

Reply:
Yeah, the problem is that the permissions need to be transitively inherited.  In that case, you'd have to keep the transitive closure of the reachable roles / channels in this attribute, which is clunky and defeats the point of using Cedar.